### PR TITLE
Ensure that we track additional member symbols in the VB binder conte…

### LINKED ISSUE #17813: BoundFieldInitializer for field initializer initializing multiple fields contains only a single field
Look at this comment for details: https://github.com/dotnet/roslyn/pull/17716#discussion_r105797433

Skipped unit test: Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.IOperationTests.MultipleFieldInitializers

```
Public Sub MultipleFieldInitializers()
            Dim source = <![CDATA[
Class C
	Dim x, y As New Object
End Class
]]>.Value
            Dim compilation = CreateCompilationWithMscorlib(source, options:=TestOptions.ReleaseDll, parseOptions:=TestOptions.Regular)

            Dim tree = compilation.SyntaxTrees.Single()
            Dim node = tree.GetRoot().DescendantNodes().OfType(Of AsNewClauseSyntax)().Single()

            compilation.VerifyOperationTree(node, expectedOperationTree:=<![CDATA[
IFieldInitializer (2 initialized fields) (OperationKind.FieldInitializerAtDeclaration)
  Field_1: C.x As System.Object
  Field_2: C.y As System.Object
  IObjectCreationExpression (Constructor: Sub System.Object..ctor()) (OperationKind.ObjectCreationExpression, Type: System.Object)
]]>.Value)
        End Sub
```

### PR DESCRIPTION
**Pull request by mavasani**
…xt and populate the additional field/property symbols initialized by VB AsNew clause that initializes more than one field/property

Note: This change also includes an API change to `IPropertyInitializer` as VB WithEvents declaration with AsNew clause initializer can initialize multiple properties.
Fixes #17813 
vso :https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/521939

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Ah, will rename the loop control variable.

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder.vb]: >field [](start = 60, length = 5)

Should doc comment mention properties as well?

File: src/Compilers/VisualBasic/Portable/Binding/DeclarationInitializerBinder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/DeclarationInitializerBinder.vb]: >additionalSymbols.All(Function(s) s.Kind = SymbolKind.Field OrElse s.Kind = SymbolKind.Property) [](start = 25, length = 96)

It feels like the assert should be stronger - 
- an empty set for symbol.Kind = SymbolKind.Parameter
- s.Kind = symbol.Kind

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder.vb]: >AdditionalContainingMembers [](start = 45, length = 27)

It feels like this method should also be overridden in 
- Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.ParametersAndLocalsBinder
- Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.EENamedTypeBinder

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: >GetInitializedFieldsOrProperties [](start = 34, length = 32)

It feels like we can do without this method.

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: >ImmutableArray.CreateRange(GetInitializedFieldsOrProperties(binder).Cast(Of FieldSymbol)) [](start = 41, length = 89)

```binder.AdditionalContainingMembers.Cast(Of FieldSymbol).Insert(0, fieldSymbol)```?

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: >ImmutableArray.CreateRange(GetInitializedFieldsOrProperties(binder).Cast(Of PropertySymbol)) [](start = 42, length = 92)

```binder.AdditionalContainingMembers.Cast(Of PropertySymbol).Insert(0, DirectCast(Me.MemberSymbol, PropertySymbol))```?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >boundPropertyInitializer.InitializedProperties.As(Of IPropertySymbol) [](start = 78, length = 69)

Should we use the same approach as CreateBoundFieldInitializerOperation?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ISymbolInitializer.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_ISymbolInitializer.vb]: >WithEvents e, f As C1 = Me'BIND:"= Me" [](start = 4, length = 38)

Please make sure we have a test for similar scenario with fields.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2).

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Actually that does not work as Cast invocation turns it into an IEnumerable, which does not have an Insert method. I am going to leave these as is for now.

---
In reply to: [149830887](https://github.com/dotnet/roslyn/pull/23077#discussion_r149830887) [](ancestors = 149830887)

**Comment by mavasani** [comment]
@AlekseyTs addressed your feedback. @cston @jcouv, can I get another compiler review?

**Comment by natidea** [comment]
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/521939

**Review by 333fred — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Binding/DeclarationInitializerBinder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/DeclarationInitializerBinder.vb]: > AndAlso (s.Kind = SymbolKind.Field OrElse s.Kind = SymbolKind.Property) [](start = 79, length = 72)

It feels like this condition is redundant, the assert below is pretty much equivalent. Can be followed up in a separate PR.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 4)

**Comment by mavasani** [comment]
@jinujoseph Are we okay to merge this with one compiler signoff and address further feedback in a separate PR?

**Comment by jinujoseph** [comment]
@mavasani , no lets wait for second review.
unless @AlekseyTs @jaredpar gives the signoff to skip second review

File: src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb
Comment (cston) [reply src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb]: Can this method simply return an `ImmutableArray(Of TSymbol)`:
`Function GetInitializedFieldsOrProperties(Of TSymbol)(binder As Binder) As ImmutableArray(Of TSymbol)`?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Binding/BinderFactory.vb
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Binding/BinderFactory.vb]: For later PR: consider asserting not null

**Review by jcouv — APPROVED** [review]
LGTM

**Pull request merged**
